### ISSUE TITLE
Repetitive setting for model, dataset,...etc

### ISSUE THREAD

**Issue by REIGN12** [ZhiyuanChen/CHANfiG#1]
Hi, I love the ideas you introduce in this repo; and in the best practice you demo in the README.md, I notice something like:
`model = Model(**config.model); optimizer=Optimizer(**config.optimizer); ...`
which is just a perfect idea.
However, there are many setting will get used by many componenet, like model, optimizer,... 
How should I handle this if using CHANfiG?

Best!

**Comment by ZhiyuanChen** [comment]
Thank you for your recognition~
Config has puzzled me for years and I didn't come up with the idea behind `CHANfiG` until very recently. 
This is also a really tricky question and will possibly take months and even years before I can find the best practice. 

I intentionally let `items()` or `**` only return value in current level, so we could have common attributes at higher level.
So for example, your `model` has an `encoder` and a `decoder`, you would like to let them shares the same droupout setting but different model size. 
```python
class ModelConfig(Config):
    def __init__(self):
        self.encoder.num_layers = 4
        self.decoder.num_layers = 2
        self.dropout = 0.1

class Model(nn.Module):
    def __init__(config):
        self.encoder = (**config.model.encoder, **config.model)
```
You can also have something like `Optimizer(abc=config.b.x, **config.optimizer)`.

**Comment by ZhiyuanChen** [comment]
Today we released the v0.0.34, which added a `Variable` class that wraps a immutable object, e.g. int, float, str to a mutable object. 
With this new feature, the same variable that is used in different position will have alternation together.

**Comment by ZhiyuanChen** [comment]
With the recent release of `v0.0.36`, we implemented `get` and `set` method of `Variable` and make `OrderedDict` calls `set` for `Variable` object. By doing so, we can have the following outputs:

```python
class Config(chanfig.Config):
    def init(self):
        emb = chanfig.Variable(384)
        self.model.emb = emb
        self.data.emb = emb

c = Config()
print(c)

Config(
  (model): Config(
    (emb): 384
  )
  (data): Config(
    (emb): 384
  )
)

c.model.emb = 256
print(d)

DC(
  (model): Config(
    (emb): 256
  )
  (data): Config(
    (emb): 256
  )
)

```

So, in your case, you could create a variable and use it everywhere.
CHANfiG will sync them automatically.
